### PR TITLE
Register layout

### PR DESCRIPTION
**Pull request by seromarin**
## Summary
- Completed register layout using reactive forms, validations with validators and custom validators, etc. 
- SCSS migrated from layout training

## Screenshot
Captura de Pantalla 2020-11-10 a la(s) 5 11 10 p  m
Captura de Pantalla 2020-11-10 a la(s) 5 11 25 p  m
Captura de Pantalla 2020-11-10 a la(s) 5 11 37 p  m
Captura de Pantalla 2020-11-10 a la(s) 5 11 57 p  m
Captura de Pantalla 2020-11-10 a la(s) 5 12 09 p  m
Captura de Pantalla 2020-11-10 a la(s) 5 13 13 p  m

## Trello Card
https://trello.com/c/2LA9NmC4/8-register-layout

### CODE REVIEW

File: wolox-books/angular.json [hunk 1]
```diff
@@ -126,6 +126,10 @@
...
-    }},
-  "defaultProject": "wolox-books"
-}
+    }
+  },
+  "defaultProject": "wolox-books",
+  "cli": {
+    "analytics": "ba969946-f146-4c73-ad86-6016808131e4"
+  }
+}
```
Comment (idcuadrowolox) [inline wolox-books/angular.json]: add empty line at the end of files

File: wolox-books/src/app/screens/register/register.component.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    this.registerForm = fb.group({
+      'name': [ null ],
+      'last-name': [ null ],
+      'email': [ null, [ Validators.required, Validators.email ] ],
+      'password': [ null, [ Validators.required, Validators.pattern(this.patternPasswordValidation) ] ],
+      'password-confirmation': [ null, [ Validators.required, passwordConfirmation() ] ],
+    })
+  }
+
+  ngOnInit(): void {
```
Comment (idcuadrowolox) [inline wolox-books/src/app/screens/register/register.component.ts]: remove empty this

File: wolox-books/src/app/screens/register/register.component.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+@Component({
+  selector: 'wb-register',
+  templateUrl: './register.component.html',
+  styleUrls: ['./register.component.scss']
+})
+export class RegisterComponent implements OnInit {
+
+  registerForm: FormGroup;
+  patternPasswordValidation: RegExp = /.*[A-Z]+.*[0-9]+.*/;
```
Comment (idcuadrowolox) [inline wolox-books/src/app/screens/register/register.component.ts]: add this to constant file

File: wolox-books/src/app/screens/register/register.component.spec.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
+import { ComponentFixture, TestBed } from '@angular/core/testing';
+
```
Comment (idcuadrowolox) [inline wolox-books/src/app/screens/register/register.component.spec.ts]: remove empty line

File: wolox-books/src/app/app.component.scss [hunk 5]
```diff
@@ -0,0 +1,10 @@
+@import './scss/equalizer';
+@import './scss/variables';
+@import './scss/breakpoints';
+@import './scss/colors';
+@import './scss/styles';
+@import './scss/form';
+@import './scss/login';
+@import './scss/navbar';
+@import './scss/detail';
+@import './scss/list';
```
Comment (idcuadrowolox) [inline wolox-books/src/app/app.component.scss]: put this files in `src/scss` folder and import them in `src/styles.scss`

Comment (idcuadrowolox) [inline wolox-books/src/app/app.component.scss]: move component style files to the respective component folder

File: wolox-books/src/app/scss/_styles.scss [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+  height: 100%;
+  justify-content: center;
+}
+
+.back-button {
+  color: $dark;
+  font-family: $primary-font;
+  font-size: 20px;
+  height: 20px;
+  margin: 50px 0 110px 0;
```
Comment (idcuadrowolox) [inline wolox-books/src/app/scss/_styles.scss]:
```suggestion
  margin: 50px 0 110px;
```

File: wolox-books/src/app/screens/register/register.component.html [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+      <input class="form-input__input" name="name" id="name" formControlName="name">
+    </div>
+    <div class="register__last-name form-input">
+      <label class="form-input__label" for="last-name">Apellido</label>
+      <input class="form-input__input" name="last-name" id="last-name" formControlName="last-name">
+    </div>
+    <div class="register__email form-input">
+      <label class="form-input__label" for="email">Email</label>
+      <input class="form-input__input" type="email" name="email" id="email" formControlName="email">
+      <span class="form-input__error" *ngIf="registerForm.controls['email']?.errors?.required && registerForm.controls['email']?.touched">El email es requerido</span>
```
Comment (idcuadrowolox) [inline wolox-books/src/app/screens/register/register.component.html]: create variable in controller for forms controls and use them here, example:
`this.emailControl = this.registerForm.controls['email'] as FormControl;`

```suggestion
      <span class="form-input__error" *ngIf="emailControl.hasError('required')">
        El email es requerido
      </span>
```

File: wolox-books/src/app/screens/register/register.component.html [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+      <span class="form-input__error" *ngIf="registerForm.controls['email']?.errors?.email && registerForm.controls['email']?.touched && registerForm.controls['email']?.value?.length > 5">Formato de email incorrecto</span>
+    </div>
+    <div class="register__password form-input">
+      <label class="form-input__label" for="password">Password</label>
+      <input class="form-input__input" type="password" name="password" id="password" formControlName="password">
+      <span class="form-input__error" *ngIf="registerForm.controls['password']?.errors?.required && registerForm.controls['password']?.touched">El password es requerido</span>
+    </div>
+    <div class="register__password-confirmation form-input">
+      <label class="form-input__label" for="password-confirmation">Confirmación de Password</label>
+      <input class="form-input__input" type="password" name="password-confirmation" id="password-confirmation" formControlName="password-confirmation">
```
Comment (idcuadrowolox) [inline wolox-books/src/app/screens/register/register.component.html]: when you have long lines it is a good idea to do this to improve readability

```suggestion
      <input
        class="form-input__input"
        type="password"
        name="password-confirmation"
        id="password-confirmation"
        formControlName="password-confirmation">
```

**Review by idcuadrowolox — CHANGES_REQUESTED** [review]

File: wolox-books/angular.json (hunk 1 above)
Comment (seromarin) [reply wolox-books/angular.json]: Added

File: wolox-books/src/app/screens/register/register.component.ts (hunk 2 above)
Comment (seromarin) [reply wolox-books/src/app/screens/register/register.component.ts]: Removed

File: wolox-books/src/app/screens/register/register.component.ts (hunk 3 above)
Comment (seromarin) [reply wolox-books/src/app/screens/register/register.component.ts]: New folder to constants and new constant file to pattern password validation

File: wolox-books/src/app/screens/register/register.component.spec.ts (hunk 4 above)
Comment (seromarin) [reply wolox-books/src/app/screens/register/register.component.spec.ts]: File removed

File: wolox-books/src/app/screens/register/register.component.html (hunk 7 above)
Comment (seromarin) [reply wolox-books/src/app/screens/register/register.component.html]: Getter created to email form control and changed to hasError() method

File: wolox-books/src/app/screens/register/register.component.html (hunk 8 above)
Comment (seromarin) [reply wolox-books/src/app/screens/register/register.component.html]: I tried...

File: wolox-books/src/app/scss/_styles.scss (hunk 6 above)
Comment (seromarin) [reply wolox-books/src/app/scss/_styles.scss]: Fixed

File: wolox-books/src/app/app.component.scss (hunk 5 above)
Comment (seromarin) [reply wolox-books/src/app/app.component.scss]: Directory structure and imports fixed

Comment (seromarin) [reply wolox-books/src/app/app.component.scss]: Clean directory structure

File: wolox-books/src/scss/_equalizer.scss [hunk 9]
```diff
@@ -0,0 +1,92 @@
+$equalizer-black: #000;
```
Comment (mcallegari10) [inline wolox-books/src/scss/_equalizer.scss]: Actually, this file is no longer needed. You can use our own equalizer https://www.npmjs.com/package/wolox-equalizer 😄

File: wolox-books/src/app/screens/register/register.component.ts [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+  registerForm: FormGroup;
+
+  constructor(
+    private fb: FormBuilder,
+  ) {
+    console.log('passwordConfirmation :>> ', passwordConfirmation);
+    this.registerForm = fb.group({
+      'first_name': [ null ],
+      'last_name': [ null ],
```
Comment (mcallegari10) [inline wolox-books/src/app/screens/register/register.component.ts:21]: I think these two fields are required

File: wolox-books/src/app/screens/register/register.component.html [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+        id="password_confirmation"
+        formControlName="password_confirmation">
+      <span
+        class="form-input__error"
+        *ngIf="passwordConfirmationFC?.hasError('required') && passwordConfirmationFC?.touched">La confirmación del password es requerida</span>
+      <span
+        class="form-input__error"
+        *ngIf="passwordConfirmationFC?.hasError('incorrectPassword') && passwordConfirmationFC?.touched && passwordConfirmationFC?.value?.length > 1">Error en la validación del password</span>
+    </div>
+    <button type="submit" class="register-button" [disabled]="!registerForm.valid">Login</button>
```
Comment (mcallegari10) [inline wolox-books/src/app/screens/register/register.component.html]:
```suggestion
    <button type="submit" class="register-button" [disabled]="!registerForm.valid">Sign Up</button>
```

File: wolox-books/src/app/screens/register/register.component.html [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+  </form>
+  <a href="./book-list.html" class="signup-button">Sign Up</a>
```
Comment (mcallegari10) [inline wolox-books/src/app/screens/register/register.component.html]:
```suggestion
  <a href="./book-list.html" class="signup-button">Login</a>
```

**Review by mcallegari10 — CHANGES_REQUESTED** [review]
Just those details, great job! 🎉

File: wolox-books/src/scss/_equalizer.scss (hunk 9 above)
Comment (seromarin) [reply wolox-books/src/scss/_equalizer.scss]: Sure! Package added and implemented!

File: wolox-books/src/app/screens/register/register.component.ts (hunk 10 above)
Comment (seromarin) [reply wolox-books/src/app/screens/register/register.component.ts:21]: image

Comment (seromarin) [reply wolox-books/src/app/screens/register/register.component.ts:21]: The card don't mention that the 'first_name' or the 'last_name' are required

File: wolox-books/src/app/screens/register/register.component.html (hunk 11 above)
Comment (seromarin) [reply wolox-books/src/app/screens/register/register.component.html]: Fixed

File: wolox-books/src/app/screens/register/register.component.html (hunk 12 above)
Comment (seromarin) [reply wolox-books/src/app/screens/register/register.component.html]: Fixed

File: wolox-books/src/app/screens/register/register.component.html (hunk 7 above)
Comment (seromarin) [reply wolox-books/src/app/screens/register/register.component.html]: Getter created to password and password_confirmation too.

File: wolox-books/src/app/screens/register/register.component.html [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+      <label class="form-input__label" for="last_name">Apellido</label>
+      <input class="form-input__input" name="last_name" id="last_name" formControlName="last_name">
+    </div>
+    <div class="register__email form-input">
+      <label class="form-input__label" for="email">Email</label>
+      <input class="form-input__input" type="email" name="email" id="email" formControlName="email">
+      <span class="form-input__error" *ngIf="emailFC?.hasError('required') && emailFC?.touched">El email es requerido</span>
+      <span
+        class="form-input__error"
+        *ngIf="emailFC?.errors?.email && emailFC?.touched && emailFC?.value?.length > 5">Formato de email incorrecto</span>
```
Comment (idcuadrowolox) [inline wolox-books/src/app/screens/register/register.component.html]:
```suggestion
        *ngIf="emailFC.hasError('email') && emailFC?.touched">
        Formato de email incorrecto
      </span>
```

**Review by idcuadrowolox — CHANGES_REQUESTED** [review]

File: wolox-books/src/app/screens/register/register.component.ts (hunk 10 above)
Comment (mcallegari10) [reply wolox-books/src/app/screens/register/register.component.ts:21]: I think all fields are required for the API to work 🤔

**Review by mcallegari10 — APPROVED** [review]
Check the required fields but I think the card with the API connection will say if those are required or not 😄

File: wolox-books/src/app/screens/register/register.component.ts (hunk 10 above)
Comment (seromarin) [reply wolox-books/src/app/screens/register/register.component.ts:21]: OK! I will make all fields required ^^

**Review by mcallegari10 — APPROVED** [review]

File: wolox-books/src/app/screens/register/register.component.html [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+        type="password"
+        name="password_confirmation"
+        id="password_confirmation"
+        formControlName="password_confirmation">
+      <span
+        class="form-input__error"
+        *ngIf="passwordConfirmationFC?.hasError('required') && passwordConfirmationFC?.touched">La confirmación del password es requerida</span>
+      <span
+        class="form-input__error"
+        *ngIf="passwordConfirmationFC?.hasError('incorrectPassword') && passwordConfirmationFC?.touched">Error en la validación del password</span>
```
Comment (idcuadrowolox) [inline wolox-books/src/app/screens/register/register.component.html]:
```suggestion
        *ngIf="passwordConfirmationFC?.hasError('incorrectPassword') && passwordConfirmationFC?.touched">
        Error en la validación del password
      </span>
```

File: wolox-books/src/app/screens/register/register.component.html [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+      <label class="form-input__label" for="password_confirmation">Confirmación de Password</label>
+      <input
+        class="form-input__input"
+        type="password"
+        name="password_confirmation"
+        id="password_confirmation"
+        formControlName="password_confirmation">
+      <span
+        class="form-input__error"
+        *ngIf="passwordConfirmationFC?.hasError('required') && passwordConfirmationFC?.touched">La confirmación del password es requerida</span>
```
Comment (idcuadrowolox) [inline wolox-books/src/app/screens/register/register.component.html]:
```suggestion
        *ngIf="passwordConfirmationFC?.hasError('required') && passwordConfirmationFC?.touched">
        La confirmación del password es requerida
      </span>
```

File: wolox-books/src/app/screens/register/register.component.html [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+    </div>
+    <div class="register__password form-input">
+      <label class="form-input__label" for="password">Password</label>
+      <input class="form-input__input" type="password" name="password" id="password" formControlName="password">
+      <span
+        class="form-input__error"
+        *ngIf="passwordFC?.hasError('required') && passwordFC?.touched">El password es requerido</span>
+      <span
+        class="form-input__error"
+        *ngIf="passwordFC?.hasError('pattern') && passwordFC?.touched">El password debe tener al menos 1 letra mayuscula y 1 numero</span>
```
Comment (idcuadrowolox) [inline wolox-books/src/app/screens/register/register.component.html]:
```suggestion
        *ngIf="passwordFC?.hasError('pattern') && passwordFC?.touched">
        El password debe tener al menos 1 letra mayuscula y 1 numero
      </span>
```

File: wolox-books/src/app/screens/register/register.component.html [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+        *ngIf="emailFC.hasError('email') && emailFC?.touched">
+        Formato de email incorrecto
+      </span>
+    </div>
+    <div class="register__password form-input">
+      <label class="form-input__label" for="password">Password</label>
+      <input class="form-input__input" type="password" name="password" id="password" formControlName="password">
+      <span
+        class="form-input__error"
+        *ngIf="passwordFC?.hasError('required') && passwordFC?.touched">El password es requerido</span>
```
Comment (idcuadrowolox) [inline wolox-books/src/app/screens/register/register.component.html]:
```suggestion
        *ngIf="passwordFC?.hasError('required') && passwordFC?.touched">
        El password es requerido
      </span>
```

File: wolox-books/src/app/screens/register/register.component.html [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+      <input class="form-input__input" name="first_name" id="first_name" formControlName="first_name">
+    </div>
+    <div class="register__last_name form-input">
+      <label class="form-input__label" for="last_name">Apellido</label>
+      <input class="form-input__input" name="last_name" id="last_name" formControlName="last_name">
+    </div>
+    <div class="register__email form-input">
+      <label class="form-input__label" for="email">Email</label>
+      <input class="form-input__input" type="email" name="email" id="email" formControlName="email">
+      <span class="form-input__error" *ngIf="emailFC?.hasError('required') && emailFC?.touched">El email es requerido</span>
```
Comment (idcuadrowolox) [inline wolox-books/src/app/screens/register/register.component.html]:
```suggestion
      <span class="form-input__error" *ngIf="emailFC?.hasError('required') && emailFC?.touched">
        El email es requerido
      </span>
```

**Review by idcuadrowolox — CHANGES_REQUESTED** [review]

**Review by idcuadrowolox — APPROVED** [review]

**Pull request merged**